### PR TITLE
Attentive encoder

### PR DESCRIPTION
**Pull request by cifkao**
Based on ["A Structured Self-attentive Sentence Embedding"](https://arxiv.org/abs/1703.03130).

### CODE REVIEW

File: neuralmonkey/encoders/attentive.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        if self.dropout_keep_prob <= 0.0 or self.dropout_keep_prob > 1.0:
+            raise ValueError("Dropout keep prob must be inside (0,1].")
+
+        with self.use_scope():
+            self.train_mode = tf.placeholder(tf.bool, [], "train_mode")
+    # pylint: enable=too-many-arguments
+
+    @tensor
+    def attention_states(self) -> tf.Tensor:
+        return dropout(get_attention_states(self.input_sequence),
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: nevim jestli to je systematický nebo ne, ale většinou se snažim to dělat tak, že se dropout volá až po vrácení (čili rule of thumb je pokud jako funkce vyrábim nějakej tenzor, vracim ho v původním stavu) - tohle by totiž pak nebyly `attention_states` ale `dropped_attention_states` nebo tak něco. Taky nevím, jestli je úplně bezpečný to po dropoutu kešovat. 

zajímavá otázka taky je, jestli chceš použít dvakrát ten samej dropout jak v `attention_weights` a v `temporal_states` nebo jestli chceš aby se to pokaždý dropoutovalo jinak. jestli ten samej, tak bych to asi nechal takhle, ale napsal bych docstring nebo tak

File: neuralmonkey/encoders/attentive.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        states = self.attention_states
+        if self.state_proj_size is not None:
+            states = tf.layers.dense(states, units=self.state_proj_size,
+                                     name="state_projection")
+
+        return tf.matmul(a=self.attention_weights, b=states, transpose_a=True)
+
+    @tensor
+    def temporal_mask(self) -> tf.Tensor:
+        return tf.ones(tf.shape(self.temporal_states)[:2], tf.float32)
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: nechceš tady spíš zkopírovat masku tý input sequence? tohle udělá všechny pozice platný.

File: neuralmonkey/encoders/attentive.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+        return output
+
+    def get_dependencies(self) -> Set[ModelPart]:
+        deps = ModelPart.get_dependencies(self)
+
+        # feed only if needed
+        if isinstance(self.input_sequence, ModelPart):
+            feedable = cast(ModelPart, self.input_sequence)
+            deps |= feedable.get_dependencies()
```
Comment (jindrahelcl) [inline neuralmonkey/encoders/attentive.py]: wow, `|=` ftw

File: neuralmonkey/encoders/attentive.py (hunk 2 above)
Comment (cifkao) [reply neuralmonkey/encoders/attentive.py]: Ne – jak je napsáno v docstringu, ty `temporal_states` mají fixní velikost a "časový rozměr" neodpovídá té vstupní sekvenci, ale jednotlivým attention heads. To je kvůli tomu, aby se pak za to dal připojit dekodér, který přes tenhle rozměr dělá další attention.

Comment (jindrahelcl) [reply neuralmonkey/encoders/attentive.py]: jo aha, už to čtu.. ok

File: neuralmonkey/encoders/attentive.py (hunk 1 above)
Comment (cifkao) [reply neuralmonkey/encoders/attentive.py]: Přijde mi, že chceme mít dropout stejnej. Ostatně koukám, že tohle je úplně stejná situace jako v [`Attention`](https://github.com/ufal/neuralmonkey/blob/master/neuralmonkey/attention/feed_forward.py#L41) (odtud jsem to asi zkopíroval). Ale už jsem z toho udělal atribut s podtržítkem.
